### ISSUE TITLE
Burn with ndarray backend is 4x slower than hardcoded model

### ISSUE THREAD

**Issue by phansel** [tracel-ai/burn#4811]
**Describe the bug**
When running a `burn-import`ed 1k parameter model under the NdArray backend on x86_64, burn is at least 10x slower than a hardcoded model, even with a substantial warmup. The outputs are identical. Am I missing something?

I started with a float32 ONNX model and imported it with burn-import. I separately converted the weights and biases (non-trained static values) to hardcoded const arrays and rebuilt the forward pass as a naïve Rust function.

The toy model torch is as follows:

```
class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.l1 = nn.Linear(6,64, bias=True)
        self.l2 = nn.Linear(64,8, bias=True)
        self.l3 = nn.Linear(8,4, bias=True)
        self.relu = nn.ReLU()

    def forward(self, x):
        x = x / torch.Tensor([7.4, 12.1, 18.5, 22.3, 81.3, 12.4])
        x = self.l1(x)
        x = self.relu(x)
        x = self.l2(x)
        x = self.relu(x)
        x = self.l3(x)
        return(x)
```
Total trainable parameters: 1004.

A typical result is a 12x slowdown with burn:
```
    Finished `release` profile [optimized + debuginfo] target(s) in 12.13s
     Running `target/release/main`
Elapsed time for burn: 6.06122ms
Elapsed time for basic matmul: 506ns
Returned value for burn: [1313.856, 1313.856, 1313.856]
Returned value for basic matmul: [1313.856, 1313.856, 1313.856, 1313.856]
```

**To Reproduce**
1. Create a simple network with Torch
2. Convert to ONNX
3. Import to burn via burn-import and burn-store
4. Convert the ONNX model to a naïve array multiplication in Rust (no imports)
5. Compare speeds

**Expected behavior**
Somewhere near 50-200% overhead. 1000% overhead seems excessive. Am I using the wrong backend?

**Desktop (please complete the following information):**
burn 0.20.1

ndarray 0.17.2

Linux 6.19 #1 SMP PREEMPT_DYNAMIC x86_64 GNU/Linux

rustc 1.93.1

**Comment by antimora** [comment]
Try with the new burn-flex backend in https://github.com/tracel-ai/burn/tree/main/crates/burn-flex

**Comment by antimora** [comment]
Also try the latest burn-onnx (burn-import was renamed) in https://github.com/tracel-ai/burn-onnx

I have added some node simplification

**Comment by phansel** [comment]
@antimora thank you for the quick response!

Updated to the HEAD of burn-flex and burn-onnx, it is down to ~2-3us per pass, or about 300-500% overhead over native.

```
    Finished `release` profile [optimized + debuginfo] target(s) in 0.20s
     Running `target/release/main`
Elapsed time for burn: 2.245839ms
Elapsed time for basic matmul: 551ns
Returned value for burn: [1313.856, 1313.856, 1313.856]
Returned value for basic matmul: [1313.856, 1313.856, 1313.856, 1313.856]
```

**Comment by antimora** [comment]
@phansel, keep in mind burn-flex is also CPU native and uses SIMD acceleration. However, the weights are on heap vs your implementation which are fixed stack based arrays. Rust compiler fully unrolls and vectorizes compile-time constants .

If you want to compare, apples to apples, try Vec based data. 

Also, realistically you should have more elements, like 2048 * 2048. burn-flex probably will beat your hand rolled model :-)

**Comment by phansel** [comment]
True, it's not apples-to-apples. 

But maybe there's room in burn-onnx for an export backend that generates `net.rs` with the compute graph inlined and parameters as constants ☺

**Comment by antimora** [comment]
> True, it's not apples-to-apples.
> 
> But maybe there's room in burn-onnx for an export backend that generates `net.rs` with the compute graph inlined and parameters as constants ☺

PRs are welcome =)

https://github.com/tracel-ai/burn-onnx

**Comment by antimora** [comment]
BTW, we will be moving away from burn-ndarray to burn-flex in the future.
